File: src/introduction/python/LibraryItem.py
# In Python, we don't need to import any specific module to use strings or booleans, 
# but for complex data structures like lists, we can import them from the collections' module.

# Base class representing a general library item
# This class has attributes for title, author, year, and checked out status
# It also has methods to check out, return, and display information about the item
# It can later be extended by specific item types like Book, Ebook, and AudioBook
# and is just a blueprint for those classes
class LibraryItem:
    def __init__(self, title, author, year):
        """
        Initializes a LibraryItem instance with the given title, author, and year.

        This method is similar to a constructor in Java and is automatically executed 
        when a new instance of the class is created. The provided arguments (title, author, year) 
        are required to initialize the object's state.

        In Python, the __init__ method handles the setup of an object's initial state, 
        just like a constructor does in Java.

        :param title: The title of the item
        :param author: The author of the item
        :param year: The year the item was published
        """

        # In Python, variables and methods are typically public by default, meaning they can be accessed from outside the class.
        # However, if we want to indicate that a variable or method should be treated as private (i.e., for internal use within the class),
        # we conventionally prefix its name with a single underscore "_". 
        # While this does not enforce true privacy, it signals to other developers that it is intended to be private.
        # Note: Python does offer a stricter form of "private," but it's less common and not conventional,
        # so it is not covered here for simplicity.
        self.title = title
        self.author = author
        self.year = year
        self.is_checked_out = False


    # In Python, we don't typically need getter and setter methods because attributes are public by default
    # and can be accessed or modified directly.

    def check_out(self):
        """
        Method to check out the item
        This method first checks if the item is already checked out, by using an if statement.
        the "not" keyword negates the value of is_checked_out.
        you can read the if statement as "if not is_checked_out do:"
        if the item is not checked out, we set is_checked_out to True and print a message.
        else (if the item is already checked out), we print a different message.
        """
        if not self.is_checked_out:
            self.is_checked_out = True
            print(f"{self.title} has been checked out.")
        else:
            print(f"{self.title} is already checked out.")

    def return_item(self):
        """
        Method to return the item
        This method is similar to the check_out method, but it checks if the item is already in the library.
        """
        if self.is_checked_out:
            self.is_checked_out = False
            print(f"{self.title} has been returned.")
        else:
            print(f"{self.title} is already in the library.")

    def display_info(self):
        """
        Method to display information about the item.
        
        Note: This method can be replaced by the __repr__ method, which provides a string 
        representation of the object and integrates better with Python’s built-in functions.
        """
        print(f"Title: {self.title}")
        print(f"Author: {self.author}")
        print(f"Year: {self.year}")
        print(f"Status: {'Checked Out' if self.is_checked_out else 'Available'}")

    def __repr__(self):
        """
        Returns a string representation of the LibraryItem object, 
        displaying the title, author, year, and checked out status.
        
        This method is more versatile and Pythonic, replacing the need for a separate 
        display_info method, as it integrates seamlessly with Python's interactive environment.

        # For example, you can simply use print(testitem) to automatically use the __repr__ method,
        # displaying a formatted string representation of the object.
        """
        status = 'Checked Out' if self.is_checked_out else 'Available'
        return f"Title: {self.title}, Author: {self.author}, Year: {self.year}, Status: {status}"

File: src/introduction/java/LibraryItem.java
package introduction.java;

// This LibraryItem is a base class representing a general library item
// This class has fields for title, author, year, and checked out status
// It also has methods to check out, return, and display information about the item
// It can later be extended by specific item types like Book, Ebook, and AudioBook
// and is just a blueprint for those classes
public class LibraryItem {
    // The keyword private, means that these fields are only accessible from inside this class,
    // or by classes that inherit from it (Like Book, Ebook, and AudioBook)
    // so this is a private field, that can only hold text values (String) and has the name title.
    private String title;
    private String author;

    // this is an integer field, that can only hold whole numbers (int) and has the name year.
    private int year;

    // this is a boolean field, that can only hold true or false values and has the name isCheckedOut.
    private boolean isCheckedOut;

    /**
     * Constructor to create a library item with title, author, and year
     * A Constructor is a special method that is called when an object is created
     * It initializes the object with the provided values (title, author, year)
     *
     * @param title The title of the item
     * @param author The author of the item
     * @param year The year the item was published
     */
    public LibraryItem(String title, String author, int year) {
        this.title = title;
        this.author = author;
        this.year = year;
        this.isCheckedOut = false;
    }

    // Getter methods to access the private fields from outside the class.
    // the public keyword means that these methods can be accessed from outside the class.
    // I hope you remember the basic syntax of a method.
    public String getTitle() {
        return title;
    }

    public String getAuthor() {
        return author;
    }

    public int getYear() {
        return year;
    }

    /**
     * Method to check out the item<hr>
     * remember that isCheckedOut is a boolean field?<br>
     * This method first checks if the item is already checked out, by using an if statement.<br>
     * the exclamation mark (!) is a logical operator that negates the value of isCheckedOut.<br>
     * so !True is False and !False is True.<br>
     * you can read the if statement as "if not isCheckedOut do:"<br>
     * if the item is not checked out, we set isCheckedOut to true and print a message.<br>
     * else (if the item is already checked out), we print a different message.<br>
     */
    public void checkOut() {
        if (!isCheckedOut) {
            isCheckedOut = true;
            System.out.println(title + " has been checked out.");
        } else {
            System.out.println(title + " is already checked out.");
        }
    }

    /**
     * Method to return the item<br>
     * This method is similar to the checkOut method, but it checks if the item is already in the library.
     */
    public void returnItem() {
        if (isCheckedOut) {
            isCheckedOut = false;
            System.out.println(title + " has been returned.");
        } else {
            System.out.println(title + " is already in the library.");
        }
    }

    /**
     * Method to display information about the item
     */
    public void displayInfo() {
        System.out.println("Title: " + title);
        System.out.println("Author: " + author);
        System.out.println("Year: " + year);
        System.out.println("Status: " + (isCheckedOut ? "Checked Out" : "Available"));
    }
}
